### PR TITLE
L1T fix int overflow

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

Addressing issue https://github.com/cms-sw/cmssw/issues/27554

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by perrotta** [comment]
Thak you @cecilecaillol for the fix.

The error reported in https://github.com/cms-sw/cmssw/issues/27554 says:
```
/build/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/a33d565f9262c93f1d0ea38c818d514b/opt/cmssw/slc7_amd64_gcc700/cms/cmssw/CMSSW_11_0_UBSAN_X_2019-07-12-2300/src/DQM/L1TMonitor/src/L1TGMT.cc:205:40: runtime error: signed integer overflow: 152916275 * 3564 cannot be represented in type 'int'
```

I am mostly worried by the orbit number 152916275 : is that a reasonable number? Shouldn't it be limited to 262144 = 2**18 orbits in 1 LS?

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
@emanueleusai you signed this, and therefore I assume that the three objections I wrote above are not deemed relevant by L1. Just for the record, could you please answer them here anyhow?

[1)](https://github.com/cms-sw/cmssw/pull/37534#issuecomment-1096366173) Is an orbit number 152916275 considered acceptable in this context? If not, shouldn't somebody investigate where such a large number comes from, and try to fix over there (perhaps there is at some point a difference between two unsigned numbers?)

[2)](https://github.com/cms-sw/cmssw/pull/37534#discussion_r848152985) Does the fix implemented at L205 work also for the firs event, when `bxnum_old_` and `obnum_old_` are still initialized at `0`?

[3)](https://github.com/cms-sw/cmssw/pull/37534#discussion_r848152985) In order to allow the conversion to `float` before assigning the right hand expression to `dBx`, did you check whether one should have included some `float` quantity already explicitely in that expression, for example `3564` --> `3564.f`?

**Comment by perrotta** [comment]
@cecilecaillol @emanueleusai any answer/comment about https://github.com/cms-sw/cmssw/pull/37534#issuecomment-1104790602?

**Comment by cecilecaillol** [comment]
@vukasinmilosevic We have this issue with the L1T DQM. It seems odd to have such large numbers in the code. Can you or someone from your team check this code makes sense and address the comments above?

**Comment by dinyar** [comment]
Hi @vukasinmilosevic,

I can have a look, however this is the DQM for the Run-1 GMT, so I don't have any special insight into that code..

Cheers,
Dinyar

**Comment by vukasinmilosevic** [comment]
Hi @dinyar, thank you for checking! I will see if there is also someone that has experience with this particular setup.

File: DQM/L1TMonitor/src/L1TGMT.cc [hunk 1]
```diff
@@ -202,12 +202,12 @@ void L1TGMT::analyze(const Event& e, const EventSetup& c) {
     //    if( (Ev - evnum_old_) == 1 && bxnum_old_ > -1 ) {
     // assume getting all events in a sequence (usefull only from reco data)
     if (bxnum_old_ > -1) {
-      int dBx = Bx - bxnum_old_ + 3564 * (e.orbitNumber() - obnum_old_);
+      float dBx = Bx - bxnum_old_ + 3564 * (e.orbitNumber() - obnum_old_);
```
[the remark this answers is not in the document]
Comment (emanueleusai) [reply DQM/L1TMonitor/src/L1TGMT.cc:205]: This issue escaped my attention, apologies. Indeed I agree with you that for the expression to be evaluated as float it should contain a float term, *before* the assignment to a float variable. @cecilecaillol  can we test if Andrea's suggestion solves the issue of large orbit numbers?

**Comment by dinyar** [comment]
> Thak you @cecilecaillol for the fix.
> 
> The error reported in #27554 says:
> 
> ```
> /build/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/a33d565f9262c93f1d0ea38c818d514b/opt/cmssw/slc7_amd64_gcc700/cms/cmssw/CMSSW_11_0_UBSAN_X_2019-07-12-2300/src/DQM/L1TMonitor/src/L1TGMT.cc:205:40: runtime error: signed integer overflow: 152916275 * 3564 cannot be represented in type 'int'
> ```
> 
> I am mostly worried by the orbit number 152916275 : is that a reasonable number? Shouldn't it be limited to 262144 = 2**18 orbits in 1 LS?

At least in hardware we only reset the orbit counter at run start. I'm not sure if this is then truncated somewhere in software though .

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1
- Tested with CMSSW_12_4_UBSAN_X_2022-04-29-1100 that this fix actually resolves the error reported for UBSAN in https://github.com/cms-sw/cmssw/issues/27554

**Pull request merged**